### DISCUSSION TITLE
Handling types for defaultValue in controlled components

### DISCUSSION THREAD

**Discussion by mcmillion** [react-hook-form/react-hook-form#9853] [Q&A] [answered]
I'm working on building some reusable components that are controlled and I'm having issues with types for `defaultValue`, which I'm assuming is due to me just not handling generics correctly. For example, my `<Switch />` component currently looks like:

```tsx
type SwitchProps<T extends FieldValues> = {
  name: Path<T>;
  control: Control<T>;
  label: string;
};

export function Switch<T extends FieldValues>(props: SwitchProps<T>) {
  const { name, control, label } = props;
  return (
    <Container>
      <Label htmlFor={name}>{label}</Label>
      <Controller
        name={name}
        control={control}
        defaultValue={false}
        render={({ field: { onChange, value, ref } }) => {
          return (
            <Root
              id={name}
              onCheckedChange={onChange}
              checked={value}
              ref={ref}
            >
              <Thumb />
            </Root>
          );
        }}
      />
    </Container>
  );
}
```

This works in practice, but the type for `defaultValue` is incorrect: `Type 'boolean' is not assignable to type 'PathValue<T, Path<T>>'.`

This works when I inline the component, so I'm assuming this is an issue with how I'm using the generic `Control<T>`, but I'm not sure how to narrow it (or constrain it) so TypeScript knows it should be boolean.

**Comment by Sajarin-M** [comment, marked as answer]
Here's what I did:
1. declared the `name` property as `FieldPathByValue<T, boolean>` , which only allows passing a boolean field as its value.
2. passed `defaultValue={false as any}`

While this may not be ideal, it at least prevents passing a non-boolean value as the name prop.

**Reply by mcmillion** [reply]
This is what I'm going with for right now, as I can at least hide the `any` part inside the component itself while keeping the public interface a bit safer and cleaner. Thanks.

**Reply by Sajarin-M** [reply]
I hope 🤞 future updates bring more efficient solutions for you.

**Comment by sanduluca** [comment]
You can't set the default value for the `name` because the type for it isn't defined yet

use the `UseControllerProps` to type the controller props. Pass the defaultValue to `Switch` component implementation
```tsx
import { FieldValues, UseControllerProps, Controller } from "react-hook-form";

type SwitchProps<T extends FieldValues> = UseControllerProps<T> & {
  label: string;
};

export function Switch<T extends FieldValues>(props: SwitchProps<T>) {
  const {name, control, label, defaultValue} = props;
  return (
    <Container>
      <Label htmlFor={name}>{label}</Label>
      <Controller
        name={name}
        control={control}
        defaultValue={defaultValue}
        render={({field: {onChange, value, ref}}) => {
          return (
            <Root
              id={name}
              onCheckedChange={onChange}
              checked={value}
              ref={ref}>
              <Thumb />
            </Root>
          );
        }}
      />
    </Container>
  );
}


function App() {
  const {control} = useForm()
  return <Switch name='like' defaultValue={false} control={control} />
}
```

**Reply by Sajarin-M** [reply]
> You can't set the default value for the `name` because the type for it isn't defined yet

Shouldn't the type of the `defaultValue` be defined as boolean if we restrict the field `name` to only boolean fields using `FieldPathByValue<T, boolean>`, but it is not the case?

**Reply by mcmillion** [reply]
This works for passing `defaultValue` in, but in this instance I'm actually trying to set the switch to default to `false` rather than `undefined` in the component itself.
